### ISSUE TITLE
The :depends keyword does not take package ownership into account

### ISSUE THREAD

**Issue by Stebalien** [syl20bnr/spacemacs#9134]
#### Description :octocat:

The layer packages `:depends` keyword does not take package ownership into
account. That is, if a layer declares a package in `<layer>-packages` but
doesn't initialize it (doesn't own it), the `:depends` keyword will still treat
the package as if it were available.

For example, given:

`layers/foo/packages.el`:

```elisp
(setq foo-packages '(foo))

(defun foo/pre-init-foo ())
```

`layers/bar/packages.el`:

```elisp
(setq bar-packages '(bar :depends foo))

(defun bar/init-bar ())
```

`bar` *will* be installed (and `foo` along with it if the `bar` package actually
depends on `foo`).

#### System Info :computer:
- OS: gnu/linux
- Emacs: 26.0.50
- Spacemacs: 0.200.9
- Spacemacs branch: local (rev. c5ef8edd)
- Graphic display: t
- Distribution: spacemacs-base
- Editing style: vim
- Completion: ivy

**Comment by TheBB** [comment]
But in this case the `bar` layer **owns** the `bar` package. The `:depends` property is checked for packages which a layer doesn't own.

**Comment by TheBB** [comment]
Just realized that it needs to do that, too.

**Comment by TheBB** [comment]
@Stebalien @skebanga Can you try https://github.com/syl20bnr/spacemacs/pull/9136

**Comment by skebanga** [comment]
First time I've done this, so not sure if it's correct, although I believe it is (followed instructions [here](https://help.github.com/articles/checking-out-pull-requests-locally/))
```
$ git fetch origin pull/9136/head:fix-9134
$ git checkout fix-9134
```
Restarted spacemacs and the warnings went away

Reverted to develop
```
$ git checkout develop
```
Restarted spacemacs and the warnings were back.

So for me, looks like it's workin! Thanks @TheBB

**Comment by Stebalien** [comment]
Hm. Maybe I'm experiencing a different issue? My problem is that helm is getting installed because:

1. The shell layer declares it but doesn't own it.
2. Several other layers pull packages that depend on helm with `(helm-something :depends helm)`.

**Comment by TheBB** [comment]
I'll solve this but it may be a few days.

**Comment by TheBB** [comment]
I have updated the PR. @Stebalien I think it should fix your problem. Please let me know.

**Comment by choppsv1** [comment]
With the latest changes I am able to re-enable ivy (disabling helm) and cleanly launch rather than failing in eieio...

**Comment by jsoo1** [comment]
I also tested on `fix-9134`.  Looks pretty good! :+1: Thanks!

**Comment by Stebalien** [comment]
Still broken for me. Describe package gives:

```
helm is an installed package.

     Status: Installed in ‘helm-20170628.2213/’ (unsigned).
    Version: 20170628.2213
    Summary: Helm is an Emacs incremental and narrowing framework
   Requires: emacs-24.4, async-1.9.2, popup-0.5.3, helm-core-2.8.0
Required by: restclient-helm-20170314.854, helm-hoogle-20161026.2234, helm-gitignore-20170210.1608, helm-css-scss-20140626.1725, helm-company-20170608.1029,
               helm-c-yasnippet-20170128.742
   Homepage: https://emacs-helm.github.io/helm/
Other versions: 20170628.2213 (melpa).
```

1. All of the "Required by" packages have `:depends helm` specified and *aren't* dependencies of any other package.
2. I don't have the helm layer enabled so nothing else should be pulling in `helm`. The only layer I have that mentions `helm` directly is the `shell` layer but that layer doesn't "own" the `helm` package.

**Comment by Stebalien** [comment]
I've tried removing the `helm` package configuration from the `shell` layer and uninstalling all helm related packages but they just get re-installed when I restart.

So, maybe this doesn't have anything to do with "unowned" packages? The one thing that fixes this for me to change the shell layer's helm dependency to:

```elisp
        (helm :toggle (configuration-layer/layer-usedp 'helm))
```

That results in a warning about the shell layer not owning the helm package but also prevents the helm package from getting installed.

However, it appears that simply *removing* the helm dependency from that package *doesn't* prevent it from being installed.

**Comment by TheBB** [comment]
Is the problem that the helm package is being downloaded or that it's being configured and enabled?

**Comment by Stebalien** [comment]
Helm is only being downloaded, not configured (although it still shouldn't be downloaded in this case). However, the other packages (the ones that declare `:depends helm`) *are* being configured (e.g., `helm-gitignore`).

**Comment by TheBB** [comment]
With the latest version of my PR I can't verify the latter problem you have. I get

    (Spacemacs) helm-gitignore is ignored since it has dependencies that are not used.

**Comment by TheBB** [comment]
@Stebalien I have updated the PR again. Please have a look.

Depending on your exact config, you will probably find that the helm package is still being installed, because the following packages depend on it (not in the Spacemacs `:depends` sense of "don't install this package if its dependencies are not present" but in the package.el sense of "to install this package, install these other ones too"):

- counsel-dash in the ivy layer depends on helm-dash which depends on helm
- emoji-cheat-sheet-plus in the emoji layer depends on helm directly
- helm-make depends on helm directly
- org-ref depends on helm-bibtex which depends on helm

 ivy layer includes the *helm-make* package, which depends on it. It's not configured to *use* helm, but since it *depends* on helm it's not possible to install it without installing helm.

**Comment by TheBB** [comment]
Sorry, I accidentally posted too soon. Here's the real comment.

@Stebalien I have updated the PR again. Please have a look.

Depending on your exact config, you will probably find that the helm package is still being installed, because the following packages (at least, there may be more) depend on it (not in the Spacemacs `:depends` sense of "don't install this package if its dependencies are not present" but in the package.el sense of "to install this package, install these other ones too"):

- counsel-dash in the ivy layer depends on helm-dash which depends on helm
- emoji-cheat-sheet-plus in the emoji layer depends on helm directly
- helm-make in the ivy layer (!) depends on helm directly
- org-ref in the bibtex layer depends on helm-bibtex which depends on helm

If you found that the helm package was not installed previously and that any of these four packages (or some I have not discovered) were used in your config, I can only assume that was a bug that has been fixed, because we cannot in good conscience override the dependencies that package authors have decided to use.

Now, whether any of these packages need to be declared with `:depends helm` or not is an open question. They were not declared with `:toggle (configuration-layer/package-usedp 'helm)` before.

**Comment by TheBB** [comment]
I'll rename `:depends` to `:requires` to avoid confusion between the two notions. :-)

**Comment by syl20bnr** [comment]
Should be fixed in develop.
@TheBB +1 for the renaming.

**Comment by Stebalien** [comment]
@TheBB The latest version (the one merged into develop) works. Thanks!

(FYI, I had already already excluded those packages for precisely this reason.)
